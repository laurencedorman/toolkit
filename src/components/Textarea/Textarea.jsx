/**
 * Copyright (c) Colibri SAS - ManoMano
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

/* eslint-disable */
// TODO in https://manomano.atlassian.net/browse/REACT-241

import React from 'react';
import PropTypes from 'prop-types';
import cn from 'classnames';
import styles from './Textarea.module.scss';

const Textarea = ({
  placeholder,
  label,
  id,
  name,
  value,
  cols,
  rows,
  required,
  disabled,
  form,
  onChange,
  className,
  dataQa,
  error,
  messageError,
  valid,
  helper,
}) => {
  const container = cn('toolkit', styles.container, className);

  const content = cn(styles.content, {
    [styles.hasValue]: value !== '',
    [styles.hasSuccess]: valid,
  });

  const textarea = cn(styles.textarea, { [styles.hasError]: error });

  const labelCn = cn({
    [styles.required]: required,
  });

  return (
    <div className={container}>
      <div className={content}>
        <textarea
          id={id}
          className={textarea}
          placeholder={placeholder}
          name={name}
          value={value}
          cols={cols}
          rows={rows}
          required={required}
          disabled={disabled}
          form={form}
          onChange={onChange}
          aria-multiline="true"
          aria-required={required}
          data-qa={dataQa}
        />
        {label && (
          <label htmlFor={id} className={labelCn}>
            {label}
          </label>
        )}
      </div>
      {helper && <span className={styles.helper}>{helper}</span>}
      {error && <span className={styles.error}>{messageError}</span>}
    </div>
  );
};

Textarea.displayName = 'Textarea';

Textarea.propTypes = {
  placeholder: PropTypes.string,
  name: PropTypes.string.isRequired,
  value: PropTypes.string,
  cols: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.element,
    PropTypes.any,
  ]),
  rows: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.element,
    PropTypes.any,
  ]),
  required: PropTypes.bool,
  disabled: PropTypes.bool,
  form: PropTypes.string,
  onChange: PropTypes.func,
  className: PropTypes.string,
  dataQa: PropTypes.string,
};

Textarea.defaultProps = {
  placeholder: '',
  label: '',
  id: '',
  value: '',
  cols: '20',
  rows: '5',
  required: false,
  disabled: false,
  form: '',
  className: '',
  dataQa: '',
  valid: false,
  error: false,
  messageError: '',
  helper: '',
};

export default Textarea;
