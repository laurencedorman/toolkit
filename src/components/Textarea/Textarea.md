```js
initialState = { value: '' };

<Textarea
  name="textarea"
  id="withLabel"
  value={state.value}
  onChange={e => setState({ value: e.target.value })}
  label="Label"
  helper="Helper data"
  required
/>;
```

```js
initialState = { value: '' };

<Textarea
  name="textarea"
  id="withLabel"
  value={state.value}
  onChange={e => setState({ value: e.target.value })}
  label="Label"
  error
  messageError="Invalid data"
/>;
```

```js
initialState = { value: '' };

<Textarea
  name="textarea"
  id="id"
  value={state.value}
  onChange={e => setState({ value: e.target.value })}
  placeholder="Placeholder"
/>;
```

```js
<Textarea name="textarea" id="id" placeholder="Disabled Placeholder" disabled />
```
