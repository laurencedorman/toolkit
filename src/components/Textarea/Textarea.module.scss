/**
 * Copyright (c) Colibri SAS - ManoMano
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

@import '../../styles/import.scss';

$tinySpace: calc(#{em($space)} / 4);

.container {
  position: relative;
  width: 100%;
  &:not(:last-child) {
    margin-bottom: em($space);
  }

  .helper,
  .error {
    display: block;
    width: 100%;
    margin: 0.25em 0 0;
    color: $greyMedDarker;
    font-size: 0.75rem;
    text-align: left;
  }

  .error {
    color: $cocoCorail;
  }
}

.content {
  position: relative;
  display: flex;
  width: 100%;

  label {
    display: block;
    position: absolute;
    top: 0;
    left: 0;
    margin: em($space) 0 0 em($space);
    color: $greyMedDarker;
    transform: translate(0) scale(1);
    transform-origin: 0 50%;
    transition: transform 0.1s linear;
    will-change: transform;

    &.required:after {
      content: '*';
      display: inline-block;
      color: $cocoCorail;
      padding-left: $tinySpace;
    }
  }

  .textarea {
    position: relative;
    width: 100%;
    padding: em($space);
    background-color: $white;
    background-clip: padding-box;
    border: em($border-width) $border-style $borderColor;
    border-radius: em($radius);
    box-sizing: border-box;
    font: em(14) $main-font;
    line-height: 1.618;
    will-change: transform;

    &::placeholder {
      display: block;
      position: absolute;
      margin-left: em($space);
      color: $greyMedDarker;
      transform: translateX(0) scale(1);
      transition: all 0.3s ease;
      opacity: 1;
      &.required:after {
        content: '*';
        display: inline-block;
        color: $cocoCorail;
        padding-left: calc(#{em($space)} / 4);
      }
    }

    &:focus::placeholder {
      transform: translateX(-100%) scale(0.5);
    }

    &:disabled {
      background-color: $greyLight;
      cursor: not-allowed;
    }
  }

  .textarea.hasError {
    border-color: $cocoCorail;
  }

  textarea:focus + label,
  &.hasValue textarea + label {
    transform: translate(em(1), em(-16)) scale(0.75);
  }
}
