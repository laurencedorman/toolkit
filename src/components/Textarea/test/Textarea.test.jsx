/**
 * Copyright (c) Colibri SAS - ManoMano
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

import React from 'react';
import { shallow } from 'enzyme';
import Textarea from '../Textarea';

describe('Textarea', () => {
  let wrapper;

  beforeEach(() => {
    wrapper = shallow(<Textarea name="name" placeholder="placeholder" />);
  });

  it('should match snapshot', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('should render a Textarea element', () => {
    expect(wrapper.find('textarea').length).toEqual(1);
  });

  it('should have no label', () => {
    expect(wrapper.find('label').length).toEqual(0);
  });

  it('should render label', () => {
    // when
    wrapper = shallow(
      <Textarea name="name" placeholder="placeholder" label="label" />
    );
    // then
    expect(wrapper.find('label').length).toEqual(1);
    expect(wrapper.find('label').text()).toEqual('label');
  });

  it('should render an helper', () => {
    // when
    wrapper = shallow(
      <Textarea name="name" placeholder="placeholder" helper="helper" />
    );
    // then
    expect(wrapper.find('span').length).toEqual(1);
    expect(wrapper.find('span').text()).toEqual('helper');
  });

  it('should render an error', () => {
    // when
    wrapper = shallow(
      <Textarea
        name="name"
        placeholder="placeholder"
        error
        messageError="error"
      />
    );
    // then
    expect(wrapper.find('span').length).toEqual(1);
    expect(wrapper.find('span').text()).toEqual('error');
  });
});
