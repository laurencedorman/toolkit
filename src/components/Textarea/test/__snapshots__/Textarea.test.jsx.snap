// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Textarea should match snapshot 1`] = `
<div
  className="toolkit container"
>
  <div
    className="content"
  >
    <textarea
      aria-multiline="true"
      aria-required={false}
      className="textarea"
      cols="20"
      data-qa=""
      disabled={false}
      form=""
      id=""
      name="name"
      placeholder="placeholder"
      required={false}
      rows="5"
      value=""
    />
  </div>
</div>
`;
